src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs

```diff
@@ -293,6 +307,57 @@ private void Init(string path, FileMode mode, FileAccess access, FileShare share
                 }
             }
 
+            // On Windows, DeleteOnClose happens when all kernel handles to the file are closed.
+            // Unix kernels don't have this feature, and .NET deletes the file when the Handle gets disposed.
+            // When the file is opened with an exclusive lock, we can use it to check the file at the path
+            // still matches the file we've opened.
+            // When the delete is performed by another .NET Handle, it holds the lock during the delete.
+            // Since we've just obtained the lock, the file will already be removed/replaced.
+            // This checks whether other Handles had DeleteOnClose for this path.
+            // To avoid performing this check for every FileStream, we only check when DeleteOnClose is set.
+            if (((options & FileOptions.DeleteOnClose) != 0) && share == FileShare.None)
```

Comment:
I think for meaningful use-cases, the user will use a mode that either opens the existing file, or creates a new one when there is no file.
So I think we can skip this check for `FileMode.CreateNew` (requires no file) and `FileMode.Open` (requires a file).